include/seqan3/range/view/single_pass_input.hpp

```diff
@@ -0,0 +1,434 @@
+
+    /*!\brief A proxy used for post-increment.
+     * \tparam value_t Effectively the `value_type_t<view_type>`.
+     *
+     * Post-increment cannot return a copy from the iterator, as it will
+     * be invalidated by the increment operation itself.
+     * Therefor a proxy is returned that caches the value of the previous position
+     * before the increment happened.
+     */
+    template <typename value_t>
+    struct single_pass_iter_proxy
+    {
+        //!\brief Stores the value before the post-increment.
+        value_t value;
+
+        //!\brief Can be used like dereferencing the iterator.
+        value_t & operator*()
+        {
+            return value;
+        }
+
+        //!\brief Implicit conversion to the value type.
+        operator value_t() const
+        {
+            return value;
+        }
+    };
+
+public:
+
+    /*!\name Member types
+     * \{
+     */
+    //!\brief Difference type.
+    using difference_type   = typename ranges::difference_type_t<base_iterator_t>;
+    //!\brief Value type.
+    using value_type        = typename ranges::value_type_t<base_iterator_t>;
+    //!\brief Pointer type.
+    using pointer           = typename std::add_pointer_t<value_type>;
+    //!\brief Reference type.
+    using reference         = typename ranges::reference_t<base_iterator_t>;
+    //!\brief Iterator category.
+    using iterator_category = std::input_iterator_tag;
+    //!\}
+
+    /*!\name Construction, destruction and assignment.
+     * \{
+     */
+    //!\brief Default construction.
+    single_pass_input_iterator() = default;
+    //!\brief Copy construction.
+    constexpr single_pass_input_iterator(single_pass_input_iterator const & rhs) = default;
+    //!\brief Move construction.
+    constexpr single_pass_input_iterator(single_pass_input_iterator && rhs) = default;
+    //!\brief Copy assignment.
+    constexpr single_pass_input_iterator & operator=(single_pass_input_iterator const & rhs) = default;
+    //!\brief Move assignment.
+    constexpr single_pass_input_iterator & operator=(single_pass_input_iterator && rhs) = default;
+    //!\brief Destruction.
+    ~single_pass_input_iterator() = default;
+
+    //!\brief Constructing from the underlying seqan3::single_pass_input_view.
+    single_pass_input_iterator(single_pass_input_view<view_type> & view) noexcept : view_ptr{&view}
+    {}
+    //!\}
+
+    /*!\name Access operations.
+     * \{
+     */
+    //!\brief Dereferences the cached iterator.
+    reference operator*() const noexcept
+    {
+        return *cached();
+    }
+    //!\}
+
+    /*!\name Operations
+     * \{
+     */
+    /*!\brief Pre-increment.
+     * \returns Reference to `*this`.
+     */
+    single_pass_input_iterator & operator++() noexcept
+    {
+        ++cached();
+        return *this;
+    }
+
+    /*!\brief Post-increment.
+     * \returns A proxy data type that stores the value before the increment.
+     */
+    auto operator++(int) noexcept
```

Comment:
have post-increment return void to avoid copying and visibly enforce single pass input_range behaviour